### DISCUSSION TITLE
Unsupported Arcitecture when using FAB_LED library

### DISCUSSION THREAD

**Discussion by ModernDayGriswolds** [SpenceKonde/megaTinyCore#328] [Library discussion]
Hello all,
I am a novice at this stuff.  I know enough to get myself into pickles like this.  Spence has messaged with me a few times about other issues, but I figured I would post on here to see if anyone else could help with this issue, as I'm sure he is a busy guy.

So, I started with the ATtiny Core and an 841 board.  My sketch compiles just fine.  Didn't like how I had to use an external crystal for 16Mhz, so I switched gears to the MegaTinyCore and an 416 chip.  No external crystal needed, but now the sketch will not compile on this Core and gives me an 'Unsupported Architecture' error.  This goes way above my head.  I understand some of the lingo, but not most.

The reason for using FAB_LED instead of the installed NeoPixel library is the RAM usage.  I need to control 700 pixels on 4 outputs.  That would require 8.4Kb of Ram minimum due to the buffers needed.  I am not writing fancy code to make the pixels dance all different colors.  It is a simple test function to turn all 700 pixels on one output to a certain color.  For function basically.  FAB_LED performs this task easily without the need for buffers so it uses hardly any RAM.

Below is the error message that I am getting.  Hopefully it's in code format.  Thanks, James

```
Arduino: 1.8.13 (Windows 10), Board: "ATtiny3216/1616/1606/816/806/416/406, No (saves space), ATtiny416, 16 MHz Internal, 1.8V (5 MHz or less), Disabled/Disabled, EEPROM retained, Enabled (default timer), UPDI (pick this unless you have an HV UPDI programmer!), 8ms, Closer to 5v"

C:\Program Files (x86)\Arduino\arduino-builder -dump-prefs -logger=machine -hardware C:\Program Files (x86)\Arduino\hardware -hardware C:\Users\James\AppData\Local\Arduino15\packages -tools C:\Program Files (x86)\Arduino\tools-builder -tools C:\Program Files (x86)\Arduino\hardware\tools\avr -tools C:\Users\James\AppData\Local\Arduino15\packages -built-in-libraries C:\Program Files (x86)\Arduino\libraries -libraries C:\Users\James\Documents\Arduino\libraries -fqbn=megaTinyCore:megaavr:atxy6:serialevent=no,chip=416,clock=16internal,bodvoltage=1v8,bodmode=disabled,eesave=enable,millis=enabled,resetpin=UPDI,startuptime=8,uartvoltage=5v -ide-version=10813 -build-path C:\Users\James\AppData\Local\Temp\arduino_build_775002 -warnings=none -build-cache C:\Users\James\AppData\Local\Temp\arduino_cache_277718 -prefs=build.warn_data_percentage=75 -prefs=runtime.tools.avrdude.path=C:\Users\James\AppData\Local\Arduino15\packages\arduino\tools\avrdude\6.3.0-arduino18 -prefs=runtime.tools.avrdude-6.3.0-arduino18.path=C:\Users\James\AppData\Local\Arduino15\packages\arduino\tools\avrdude\6.3.0-arduino18 -prefs=runtime.tools.avr-gcc.path=C:\Users\James\AppData\Local\Arduino15\packages\DxCore\tools\avr-gcc\7.3.0-atmel3.6.1-azduino3 -prefs=runtime.tools.avr-gcc-7.3.0-atmel3.6.1-azduino3.path=C:\Users\James\AppData\Local\Arduino15\packages\DxCore\tools\avr-gcc\7.3.0-atmel3.6.1-azduino3 -prefs=runtime.tools.python3.path=C:\Users\James\AppData\Local\Arduino15\packages\megaTinyCore\tools\python3\3.7.2-post1 -prefs=runtime.tools.python3-3.7.2-post1.path=C:\Users\James\AppData\Local\Arduino15\packages\megaTinyCore\tools\python3\3.7.2-post1 -prefs=runtime.tools.arduinoOTA.path=C:\Users\James\AppData\Local\Arduino15\packages\arduino\tools\arduinoOTA\1.3.0 -prefs=runtime.tools.arduinoOTA-1.3.0.path=C:\Users\James\AppData\Local\Arduino15\packages\arduino\tools\arduinoOTA\1.3.0 -verbose C:\Users\James\Documents\Arduino\LED_tests\LED_tests.ino

C:\Program Files (x86)\Arduino\arduino-builder -compile -logger=machine -hardware C:\Program Files (x86)\Arduino\hardware -hardware C:\Users\James\AppData\Local\Arduino15\packages -tools C:\Program Files (x86)\Arduino\tools-builder -tools C:\Program Files (x86)\Arduino\hardware\tools\avr -tools C:\Users\James\AppData\Local\Arduino15\packages -built-in-libraries C:\Program Files (x86)\Arduino\libraries -libraries C:\Users\James\Documents\Arduino\libraries -fqbn=megaTinyCore:megaavr:atxy6:serialevent=no,chip=416,clock=16internal,bodvoltage=1v8,bodmode=disabled,eesave=enable,millis=enabled,resetpin=UPDI,startuptime=8,uartvoltage=5v -ide-version=10813 -build-path C:\Users\James\AppData\Local\Temp\arduino_build_775002 -warnings=none -build-cache C:\Users\James\AppData\Local\Temp\arduino_cache_277718 -prefs=build.warn_data_percentage=75 -prefs=runtime.tools.avrdude.path=C:\Users\James\AppData\Local\Arduino15\packages\arduino\tools\avrdude\6.3.0-arduino18 -prefs=runtime.tools.avrdude-6.3.0-arduino18.path=C:\Users\James\AppData\Local\Arduino15\packages\arduino\tools\avrdude\6.3.0-arduino18 -prefs=runtime.tools.avr-gcc.path=C:\Users\James\AppData\Local\Arduino15\packages\DxCore\tools\avr-gcc\7.3.0-atmel3.6.1-azduino3 -prefs=runtime.tools.avr-gcc-7.3.0-atmel3.6.1-azduino3.path=C:\Users\James\AppData\Local\Arduino15\packages\DxCore\tools\avr-gcc\7.3.0-atmel3.6.1-azduino3 -prefs=runtime.tools.python3.path=C:\Users\James\AppData\Local\Arduino15\packages\megaTinyCore\tools\python3\3.7.2-post1 -prefs=runtime.tools.python3-3.7.2-post1.path=C:\Users\James\AppData\Local\Arduino15\packages\megaTinyCore\tools\python3\3.7.2-post1 -prefs=runtime.tools.arduinoOTA.path=C:\Users\James\AppData\Local\Arduino15\packages\arduino\tools\arduinoOTA\1.3.0 -prefs=runtime.tools.arduinoOTA-1.3.0.path=C:\Users\James\AppData\Local\Arduino15\packages\arduino\tools\arduinoOTA\1.3.0 -verbose C:\Users\James\Documents\Arduino\LED_tests\LED_tests.ino

Using board 'atxy6' from platform in folder: C:\Users\James\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.2.7

Using core 'megatinycore' from platform in folder: C:\Users\James\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.2.7

Detecting libraries used...

"C:\\Users\\James\\AppData\\Local\\Arduino15\\packages\\DxCore\\tools\\avr-gcc\\7.3.0-atmel3.6.1-azduino3/bin/avr-g++" -c -g -Os -w -std=gnu++11 -fpermissive -fno-exceptions -ffunction-sections -fdata-sections -fno-threadsafe-statics -Wno-error=narrowing -flto -w -x c++ -E -CC -mmcu=attiny416 -DF_CPU=16000000L -DCLOCK_SOURCE=0 -DARDUINO=10813 -DARDUINO_AVR_ATtiny416 -DARDUINO_ARCH_MEGAAVR "-DMEGATINYCORE=\"2.2.7\"" -DMEGATINYCORE_MAJOR=2UL -DMEGATINYCORE_MINOR=2UL -DMEGATINYCORE_PATCH=7UL -DMEGATINYCORE_RELEASED=1 -DARDUINO_attinyxy6 -DNOSERIALEVENT -DMILLIS_USE_TIMERD0 -DUARTBAUD5V "-IC:\\Users\\James\\AppData\\Local\\Arduino15\\packages\\megaTinyCore\\hardware\\megaavr\\2.2.7\\cores\\megatinycore/api/deprecated" "-IC:\\Users\\James\\AppData\\Local\\Arduino15\\packages\\megaTinyCore\\hardware\\megaavr\\2.2.7\\cores\\megatinycore" "-IC:\\Users\\James\\AppData\\Local\\Arduino15\\packages\\megaTinyCore\\hardware\\megaavr\\2.2.7\\variants\\txy6" "C:\\Users\\James\\AppData\\Local\\Temp\\arduino_build_775002\\sketch\\LED_tests.ino.cpp" -o nul -DARDUINO_LIB_DISCOVERY_PHASE

Alternatives for FAB_LED.h: [FAB_LED-master@1.0.0]

ResolveLibrary(FAB_LED.h)

  -> candidates: [FAB_LED-master@1.0.0]

"C:\\Users\\James\\AppData\\Local\\Arduino15\\packages\\DxCore\\tools\\avr-gcc\\7.3.0-atmel3.6.1-azduino3/bin/avr-g++" -c -g -Os -w -std=gnu++11 -fpermissive -fno-exceptions -ffunction-sections -fdata-sections -fno-threadsafe-statics -Wno-error=narrowing -flto -w -x c++ -E -CC -mmcu=attiny416 -DF_CPU=16000000L -DCLOCK_SOURCE=0 -DARDUINO=10813 -DARDUINO_AVR_ATtiny416 -DARDUINO_ARCH_MEGAAVR "-DMEGATINYCORE=\"2.2.7\"" -DMEGATINYCORE_MAJOR=2UL -DMEGATINYCORE_MINOR=2UL -DMEGATINYCORE_PATCH=7UL -DMEGATINYCORE_RELEASED=1 -DARDUINO_attinyxy6 -DNOSERIALEVENT -DMILLIS_USE_TIMERD0 -DUARTBAUD5V "-IC:\\Users\\James\\AppData\\Local\\Arduino15\\packages\\megaTinyCore\\hardware\\megaavr\\2.2.7\\cores\\megatinycore/api/deprecated" "-IC:\\Users\\James\\AppData\\Local\\Arduino15\\packages\\megaTinyCore\\hardware\\megaavr\\2.2.7\\cores\\megatinycore" "-IC:\\Users\\James\\AppData\\Local\\Arduino15\\packages\\megaTinyCore\\hardware\\megaavr\\2.2.7\\variants\\txy6" "-IC:\\Users\\James\\Documents\\Arduino\\libraries\\FAB_LED-master" "C:\\Users\\James\\AppData\\Local\\Temp\\arduino_build_775002\\sketch\\LED_tests.ino.cpp" -o nul -DARDUINO_LIB_DISCOVERY_PHASE

Error while detecting libraries included by C:\Users\James\AppData\Local\Temp\arduino_build_775002\sketch\LED_tests.ino.cpp

Generating function prototypes...

"C:\\Users\\James\\AppData\\Local\\Arduino15\\packages\\DxCore\\tools\\avr-gcc\\7.3.0-atmel3.6.1-azduino3/bin/avr-g++" -c -g -Os -w -std=gnu++11 -fpermissive -fno-exceptions -ffunction-sections -fdata-sections -fno-threadsafe-statics -Wno-error=narrowing -flto -w -x c++ -E -CC -mmcu=attiny416 -DF_CPU=16000000L -DCLOCK_SOURCE=0 -DARDUINO=10813 -DARDUINO_AVR_ATtiny416 -DARDUINO_ARCH_MEGAAVR "-DMEGATINYCORE=\"2.2.7\"" -DMEGATINYCORE_MAJOR=2UL -DMEGATINYCORE_MINOR=2UL -DMEGATINYCORE_PATCH=7UL -DMEGATINYCORE_RELEASED=1 -DARDUINO_attinyxy6 -DNOSERIALEVENT -DMILLIS_USE_TIMERD0 -DUARTBAUD5V "-IC:\\Users\\James\\AppData\\Local\\Arduino15\\packages\\megaTinyCore\\hardware\\megaavr\\2.2.7\\cores\\megatinycore/api/deprecated" "-IC:\\Users\\James\\AppData\\Local\\Arduino15\\packages\\megaTinyCore\\hardware\\megaavr\\2.2.7\\cores\\megatinycore" "-IC:\\Users\\James\\AppData\\Local\\Arduino15\\packages\\megaTinyCore\\hardware\\megaavr\\2.2.7\\variants\\txy6" "-IC:\\Users\\James\\Documents\\Arduino\\libraries\\FAB_LED-master" "C:\\Users\\James\\AppData\\Local\\Temp\\arduino_build_775002\\sketch\\LED_tests.ino.cpp" -o "C:\\Users\\James\\AppData\\Local\\Temp\\arduino_build_775002\\preproc\\ctags_target_for_gcc_minus_e.cpp" -DARDUINO_LIB_DISCOVERY_PHASE

In file included from C:\Users\James\Documents\Arduino\LED_tests\LED_tests.ino:1:0:

C:\Users\James\Documents\Arduino\libraries\FAB_LED-master/FAB_LED.h:350:2: error: #error "Unsupported Architecture"

 #error "Unsupported Architecture"

  ^~~~~

Using library FAB_LED-master at version 1.0.0 in folder: C:\Users\James\Documents\Arduino\libraries\FAB_LED-master 

exit status 1

Error compiling for board ATtiny3216/1616/1606/816/806/416/406.
```

**Comment by SpenceKonde** [comment]
Sorry for not getting back sooner to your PM - havcen;t had a chance to look into it until now... 

Oh dear....  Well, yes, that does appear to be accurate - he has a bunch of architecture-specific macros for different types of processors, and the classic AVRs are ARDUINO_ARCH_AVR, these are ARDUINO_ARCH_MEGAAVR, and the timing of relevant instructions *has* changed on these parts (this guy's code relies on cbi/sbi, which are single cycle instead of 2 cycle now)

I am once more in awe of that dude. Are the requirements for supporting a new architecture obscure? Undocumented? scattered across many files? Or are they clearly described in one place? 

Try https://github.com/SpenceKonde/FAB_LED/blob/master/FAB_LED.h 

It compiles now (haven[t tested with LEDs), if you confirm it works I'll send him a PR. 

As an aside, euhm..... on the 841 you pretty much can run at 16 MHz from the internal oscillator. I think I had one chip out of the dozen or so I tested with that had to be cranked all the way to 254 or 255? Current version supports this, but takes a guess on the new OSCCAL0, Future version of ATTinyCore (probably the next release) will provide proper tuning support (I needed the tooling for it written anyway). The oscillator on the 841/441 is unique, wonderful, and infuriating (the strong voltage dependence - which does help for 16@5v)

**Comment by ModernDayGriswolds** [comment]
Its all good.  Beggars can't be choosers when asking for help.  I appreciate you doing this for the community.  I figured I would move our conversation here in the obscure chance that someone else may run into an issue with FAB_LED.  Unlikely, but you never know.

I really like the ease of programming with the UPDI port on the MegaTinyCore, so that was an added benefit of using the 416 over the 841.

I'm still having issues with compiling now though.....  I originally installed just your FAB_LED.h file.  That didn't compile so I then completely deleted the original FAB_LED and installed yours.  Got the same error message below.  Is there any different settings I need to have selected associated with the board?  I wouldn't think so.
![image](https://user-images.githubusercontent.com/52804053/108300399-e0acf900-7165-11eb-8cc9-7944a1e69e57.png)

```
c:/users/james/appdata/local/arduino15/packages/dxcore/tools/avr-gcc/7.3.0-atmel3.6.1-azduino3/bin/../lib/gcc/avr/7.3.0/../../../../avr/bin/ld.exe: region `text' overflowed by 1701 bytes
```

**Reply by ModernDayGriswolds** [reply]
Just checked the other chips. It seems to compile on every chip except the 406 and the 416.
I have an 816 here.  I will solder that one up and see if it works or not.
James

**Reply by ModernDayGriswolds** [reply]
Was able to upload it to an 816, but no led's functioning.  I will upload it to my Pro mini tomorrow to verify that my code is still functioning as it should to make the leds illuminate.  I know there is a way to get serial monitor to work on the 816 with the TTL, but that is not working for me.  Can I use my FTDI Friend from AdaFruit to get serial communications?
Thanks,
James

**Comment by SpenceKonde** [comment]
That's the wacky error you get when the sketch compiles to something too big to fit in the flash of the selected chip, and that chip is smaller than 8k  - the combination of having rjmp, not jmp and memory mapped flash leaves the compiler giving these ugly messages instead of a normal one saying it doesn't fit. The useful part of that error is: `ld.exe: region 'text' overflowed by 1701 bytes`

**Reply by ModernDayGriswolds** [reply]
Spence,
So I uploaded the program code to my Pro Mini, only changing the input and output addresses, with your FAB_LED.h file still in there and it compiled and worked just fine.  So there is something else going on between the core and the FAB_LED library that is not jiving.  
Can I bother to ask you if there is anything else that can be done to get it to work??
Thanks,
James

**Comment by SpenceKonde** [comment]
Nothing I changed would be compiled in if built for a pro mini - that's
classic AVR, ARDUINO_ARCH_AVR, the changes I made are only pulled in for
ARDUINO_ARCH_MEGAAVR devices, like the ones supported by megaTinyCore.

On Thu, Feb 18, 2021 at 9:24 PM ModernDayGriswolds <notifications@github.com>
wrote:

> Spence,
> So I uploaded the program code to my Pro Mini, only changing the input and
> output addresses, with your FAB_LED.h file still in there and it compiled
> and worked just fine. So there is something else going on between the core
> and the FAB_LED library that is not jiving.
> Can I bother to ask you if there is anything else that can be done to get
> it to work??
> Thanks,
> James
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/SpenceKonde/megaTinyCore/discussions/328#discussioncomment-382926>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTXEW2BP65QWDUHG3VLXVTS7XDXTANCNFSM4XZHH62A>
> .
>


-- 
____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore <https://github.com/SpenceKonde/ATTinyCore>: Arduino support for
all pre-2016 tinyAVR with >2k flash!
megaTinyCore <https://github.com/SpenceKonde/megaTinyCore>: Arduino support
for all post-2016 tinyAVR parts!
DxCore <https://github.com/SpenceKonde/DxCore>: Arduino support for the AVR
Dx-series parts, the latest and greatest from Microchip!
Contact: spencekonde@gmail.com

**Reply by ModernDayGriswolds** [reply]
Ah, gotcha.  So is there anything else that I could try?

**Comment by SpenceKonde** [comment]
The one thing I'd try is.. pick a pin in port A where the same-numbered pin
in PORTB is also exposed. Check that pin too....I'm not clear where the
port ID is coming from...
If they're getting it from core-supplied macrolike digital pin to bit mask,
but hardcode A=1, B=2 ... they''d be off by 1 port, because PORTA is
numbered 0 on modern AVR, and 1 on classic AVR. (classic AVR uses 0 for
NOT_A_PORT, modern AVR uses -1 or 255, for NOT_A_PORT)

Also, do you have a 'scope? It would be worth checking to see if data is
coming out of some pin, just nothing valid enough to control pixels. I find
it sort of hard to believe that it could compile but do *nothing* with all
his code behind it, so what's going on?

On Thu, Feb 18, 2021 at 9:40 PM ModernDayGriswolds <notifications@github.com>
wrote:

> Ah, gotcha. So is there anything else that I could try?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/SpenceKonde/megaTinyCore/discussions/328#discussioncomment-382966>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTXEW6GS4SHO2UC6C7JRMLS7XFSRANCNFSM4XZHH62A>
> .
>


-- 
____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore <https://github.com/SpenceKonde/ATTinyCore>: Arduino support for
all pre-2016 tinyAVR with >2k flash!
megaTinyCore <https://github.com/SpenceKonde/megaTinyCore>: Arduino support
for all post-2016 tinyAVR parts!
DxCore <https://github.com/SpenceKonde/DxCore>: Arduino support for the AVR
Dx-series parts, the latest and greatest from Microchip!
Contact: spencekonde@gmail.com

**Reply by ModernDayGriswolds** [reply]
OK, So got some time in tonight on it.  my LED pins for the ws2811 pixels outputs are as follows:
```
const int Led1 = PIN_PB0;
const int Led2 = PIN_PC0;
const int Led3 = PIN_PA4;
const int Led4 = PIN_PA6;
```
With this code running, I took the data pin around to all open pins on the board.  Only one started working. That is PIN_PA2.  Pin 15 on arduino.  Physical pin 18 on the 816.

So I changed the Led# assignments over to Arduino pin numbers.  Same pins as before in the PIN_Pnx but with Arduino numbers in place:
```
const int Led1 = 9;
const int Led2 = 10;
const int Led3 = 0;
const int Led4 = 2;
```
Got the same result.  Only one worked. That is PIN_PA2.  Pin 15 on arduino.  Physical pin 18 on the 816.

So now I changed the pins to different Arduino pin numbers as follows:
```
const int Led1 = 1;
const int Led2 = 2;
const int Led3 = 3;
const int Led4 = 4;
```
Now this time I got a different result.  This thoroughly  confuses me.  I have four outputs that function, but none coincides with what is supposed to be working.  The working Arduino pins now are: 0,14,15,16.  

Again, changed to different Arduino numbers:
```
const int Led1 = 5;
const int Led2 = 6;
const int Led3 = 7;
const int Led4 = 8;
```
Again, got a different result.  Now I only have three working outputs.  Arduino pins: 1,2,3.  That is it.

Can you make heads or tails of this?  Where are the numbers assignments coming from?
I plan to do more testing by making a program that only has one output and see if I can't get a mapping of what number is what on this 816.

So, yeah, you were right, it is spitting out data, but just not where I would expect it.  Plus side though, it is working!!
James

**Comment by SpenceKonde** [comment]
Yes actually, that does make sense.... the number you're passing is being
used as the bit of the pin on port A. How do you tell the library which
pins to use? When I looked at the examples, it looked like they had a weird
way of specifying it, by port and bit, not arduino pin number. Are you
using that correctly? I'm not sure that that library attempts to convert
arduino pin numbers...  the examples I saw had it spec'ed by port and bit.
If it was defaulting port a and using what was passed as bit, that is what
you would see

____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: spencekonde@gmail.com

**Comment by SpenceKonde** [comment]
See the pinout chart in the repo, 0~7 gets you output on PA0~PA7. PA0 is
the updi pin, though, so you wont get output there.

____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: spencekonde@gmail.com

**Reply by ModernDayGriswolds** [reply]
Once again, you are correct.  I was still using the example architecture and had D as the port.  816 does not have a D port, so it was reverting to the A port.  In the process of trying to make the program streamline with integers, I forgot about the port assignment on the FAB_LED side.  Now everything seems to be functioning fine!!
Greatly appreciate your help on this.  Thank you.
Please shoot me your PayPal, I'd like to buy you a beer sir. 
James

**Reply by SpenceKonde** [reply]
Great! I sent him a PR, but I have a feeling he will ignore it like he is the other PRs and issues :(
